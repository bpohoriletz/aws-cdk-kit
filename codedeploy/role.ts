import * as cdk from "aws-cdk-lib";
import * as iam from "aws-cdk-lib/aws-iam";

import * as con from "../utils/naming";

// TOFIX; delete
export function createCodedeployRole(resourceNamePrefix: string[], regionalCdBucketArn: string,  stack: cdk.Stack) : iam.Role {
  const resourceName = con.roleName(resourceNamePrefix.concat("codedeploy"));

  const role = new iam.Role(stack, resourceName, {
    assumedBy: new iam.ServicePrincipal("codedeploy.amazonaws.com"),
    managedPolicies: [
      iam.ManagedPolicy.fromAwsManagedPolicyName("service-role/AWSCodeDeployRole"),
    ],
    roleName: resourceName,
  });
  const allowEc2UploadToEbRegionalCdBucket: iam.Policy = new iam.Policy(stack, "allowEc2UploadToEbRegionalCdBucket", {
    statements: [
      new iam.PolicyStatement({
        actions: ["s3:PutObject", "s3:ListBucket", "s3:ListBucketVersions", "s3:GetObject", "s3:GetObjectVersion"],
        resources: [regionalCdBucketArn, `${regionalCdBucketArn}/resources/*`],
      }),
    ],
  });
  allowEc2UploadToEbRegionalCdBucket.attachToRole(role);

  return role;
};
