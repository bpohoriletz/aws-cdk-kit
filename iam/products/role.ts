import { Stack } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';

export class RoleProduct implements IRoleProduct {
  assumedBy: iam.IPrincipal;
  description?: string;
}

interface IRoleProduct extends iam.RoleProps {
  assumedBy: iam.IPrincipal;
  description?: string;
}

export interface IRoleBuilder {
  setInlinePolicies(resources: string[]): this;
  setManagedPolicies(): this;
  getResult(): iam.Role;
}

export interface IRoleBuilderConstructor {
  new (stack: Stack, id: string): IRoleBuilder;
}
