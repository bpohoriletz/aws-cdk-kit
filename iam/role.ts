import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import PolicyStatementDirector from '../directors/policy-statement';
import S3PolicyStatementBuilder from './policy-statement-builders/s3';

export function createGithubCliRole(githubAccount: string, stack: cdk.Stack): iam.Role {
  const name = 'GitHubActionsRole';
  new iam.OpenIdConnectProvider(stack, 'GitHubOIDCProvider', {
    url: 'https://token.actions.githubusercontent.com',
    clientIds: ['sts.amazonaws.com'],
    thumbprints: ['6938fd4d98bab03faadb97b34396831e3780aea1'], // GitHub's current root CA thumbprint
  });
  const cliRole = new iam.Role(stack, name, {
    assumedBy: new iam.FederatedPrincipal(
      `arn:aws:iam::${stack.account}:oidc-provider/token.actions.githubusercontent.com`,
      {
        StringEquals: {
          'token.actions.githubusercontent.com:aud': 'sts.amazonaws.com',
        },
        StringLike: {
          'token.actions.githubusercontent.com:sub': `repo:${githubAccount}/*`,
        },
      },
      'sts:AssumeRoleWithWebIdentity'
    ),
    roleName: name,
    description: 'Role for GitHub Actions to access AWS resources',
  });
  cliRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AWSElasticBeanstalkWebTier'));
  cliRole.addManagedPolicy(
    iam.ManagedPolicy.fromAwsManagedPolicyName('AWSElasticBeanstalkManagedUpdatesCustomerRolePolicy')
  );
  cliRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('SecretsManagerReadWrite'));
  const resources = ['arn:aws:s3:::codedeploy-*', 'arn:aws:s3:::codedeploy-*/*'];
  new iam.Policy(stack, 'accessCodeDeployBucket', {
    statements: [new PolicyStatementDirector(S3PolicyStatementBuilder).constructS3ReadWritePolicyStatement(resources)],
  }).attachToRole(cliRole);
  const codeDeploy: iam.Policy = new iam.Policy(stack, 'codeDeploy', {
    statements: [
      new iam.PolicyStatement({
        actions: ['codedeploy:*'],
        resources: ['*'],
      }),
    ],
  });
  codeDeploy.attachToRole(cliRole);
  new cdk.CfnOutput(stack, 'GitHub Actions Role ARN', {
    value: cliRole.roleArn,
    description: 'The ARN of the IAM Role that should be assumed in GitHub',
  });

  return cliRole;
}
