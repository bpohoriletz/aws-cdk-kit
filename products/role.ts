import * as iam from 'aws-cdk-lib/aws-iam';

export class RoleProduct implements iam.RoleProps {
  assumedBy: iam.IPrincipal;
  description: string;

  setAssumedBy(principal: iam.IPrincipal): void {
    this.assumedBy = principal;
  }

  setDescription(description: string): void {
    this.description = description;
  }
}

export interface IRoleBuilder {
  getResult(): RoleProduct;
}

export interface IRoleBuilderConstructor {
  new (): IRoleBuilder;
}
