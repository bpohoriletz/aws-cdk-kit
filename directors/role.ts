import * as iam from 'aws-cdk-lib/aws-iam';
import { IRoleBuilderConstructor } from '../products/role';
import { Stack } from 'aws-cdk-lib';

export default class RoleDirector {
  private builder: IRoleBuilderConstructor;

  constructor(builder: IRoleBuilderConstructor) {
    this.builder = builder;
  }

  constructEc2Role(stack: Stack, roleId: string, bucketArn: string): iam.Role {
    return new this.builder(stack, roleId).setInlinePolicies([bucketArn]).setManagedPolicies().getResult();
  }
}
