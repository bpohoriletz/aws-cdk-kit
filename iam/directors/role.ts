import * as iam from 'aws-cdk-lib/aws-iam';
import { IRoleBuilderConstructor } from '../products/role';
import { Stack } from 'aws-cdk-lib';

export default class RoleDirector {
  private builder: IRoleBuilderConstructor;

  constructor(builder: IRoleBuilderConstructor) {
    this.builder = builder;
  }

  constructEc2Role(stack: Stack, roleId: string, ec2BucketArn: string): iam.Role {
    return new this.builder(stack, roleId).setInlinePolicies([ec2BucketArn]).setManagedPolicies().getResult();
  }

  constructGtihubRole(stack: Stack, accountName: string): iam.Role {
    return new this.builder(stack, accountName).setInlinePolicies([]).setManagedPolicies().getResult();
  }
}
