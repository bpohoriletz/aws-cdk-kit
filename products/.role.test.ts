import { RoleProduct } from './role';

describe('new RoleProduct()', () => {
  let instance: RoleProduct;

  beforeEach(() => {
    instance = new RoleProduct();
  });

  test('has correct defaults', () => {
    expect(instance).toMatchSnapshot();
  });
});
