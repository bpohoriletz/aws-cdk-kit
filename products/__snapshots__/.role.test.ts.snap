// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new RoleProduct() has correct defaults 1`] = `RoleProduct {}`;
